src/tools/concurrency/TraceParser.java

```diff
@@ -0,0 +1,342 @@
+  }
+
+  private void parseMessage(final byte type, final Queue<Long> createdPromises) {
+    long promid = 0;
+    // promise msg
+    if ((type & PROMISE_BIT) > 0) {
+      promid = b.getLong(); // promise
+    }
+
+    long sender = b.getLong(); // sender
+    b.getLong(); // causal message
+    short sym = b.getShort(); // selector
+
+    if (!expectedMessages.containsKey(currentReceiver)) {
+      expectedMessages.put(currentReceiver, new PriorityQueue<>());
+    }
+
+    if ((type & PROMISE_BIT) > 0) {
+      expectedMessages.get(currentReceiver).add(new PromiseMessage(sender, symbolMapping.get(sym), promid, currentMailbox, msgNo, createdPromises));
+    } else {
+      expectedMessages.get(currentReceiver).add(new Message(sender, symbolMapping.get(sym), currentMailbox, msgNo, createdPromises));
+    }
+
+    // timestamp
+    if ((type & TIMESTAMP_BIT) > 0) {
+      b.getLong();
+      b.getLong();
+    }
+
+    // params
+    if ((type & PARAMETER_BIT) > 0) {
+      byte numParam = b.get();
+
+      for (int i = 0; i < numParam; i++) {
+        parseParameter();
+      }
+    }
+    currentMessage++;
+    msgNo++;
+
+  }
+
+  private void parseParameter() {
+    byte type = b.get();
+    switch (ActorExecutionTrace.ParamTypes.values()[type]) {
+      case False:
+        break;
+      case True:
+        break;
+      case Long:
+        b.getLong();
+        break;
+      case Double:
+        b.getDouble();
+        break;
+      case Promise:
+        b.getLong();
+        break;
+      case Resolver:
+        b.getLong();
+        break;
+      case Object:
+        b.getShort();
+        break;
+      case String:
+        break;
+      default:
+        break;
+    }
+  }
+
+  private static class ActorNode implements Comparable<ActorNode> {
+    final long actorId;
+    int childNo;
+    int mailboxNo;
+    ArrayList<ActorNode> children;
+
+    ActorNode(final long actorId) {
+      super();
+      this.actorId = actorId;
+    }
+
+    @Override
+    public int compareTo(final ActorNode o) {
+      int i = Integer.compare(mailboxNo, o.mailboxNo);
+      if (i == 0) {
+        i = Integer.compare(childNo, o.childNo);
+      }
+      if (i == 0) {
+        i = Long.compare(actorId, o.actorId);
+      }
+
+      return i;
+    }
+
+    private void addChild(final ActorNode child) {
+      if (children == null) {
+        children = new ArrayList<>();
+      }
+
+      child.childNo = children.size();
+      children.add(child);
+      java.util.Collections.sort(children);
+    }
+
+    protected ActorNode getChild(final int childNo) {
+      assert children != null : "Actor does not exist in trace!";
+      assert children.size() > childNo : "Actor does not exist in trace!";
+      return children.get(childNo);
+    }
+
+    @Override
+    public String toString() {
+      return "" + actorId + ":" + childNo;
+    }
+  }
+
+  public static class Message implements Comparable<Message> {
```

Comment:
Could we use another name than a plain 'Message' here?
Don't know, perhaps MessageRecord?